watidl/watidl.md

```diff
@@ -0,0 +1,119 @@
+# WatIDL :duck:
+
+
+WatIDL (`/ˈwädl/`) is an IDL for webassembly. Its goal is to be able to fully describe the import and export capabilities of Webassembly modules while being intuitive to read and reason about. It inherits the Webassembly text format's syntax and attempts to closely follow and fit into the Webassembly ecosystem. WatIDL should used as a simple syntax for describing the interfaces along with WebIDL bindings to describe how the different data structures are formed and received by Wasm modules.
+
+# Differences from WAT
+While WatIDL follow close to WAT there are some differences
+
+## Types
+WatIDL uses the following primitive types
+
+```
+intType := "i8" | "i6" | "i32" | "i64" | "u8" | "u16" | "u32" | "u64";
+floatType ::= "f32" | "f64" 
+primType ::= "null" | "bool" | "string" | "data" | intTypes | floatType;
+```
+
+In addition new types can also be created using
+- struct
+- array
+- union
+- enum
+
+The syntax for these follows the [gc proposals](https://github.com/WebAssembly/gc/blob/master/proposals/gc/Overview.md) syntax when possible.
+
+```
+(type $drinks
+  (enum
+    $coffee
+    $juice
+    $water
+    $wine
+  )
+)
+
+(type $baz
+  (struct
+    (field $foo $drinks)
+    (field $bar bool)
+  )
+)
+```
+
+## Interface and Functions
+### Imports and Exports
+Instead of having "module"s WatIDL has "interfaces" which has nearly the same syntax as [modules](https://webassembly.github.io/spec/core/text/modules.html). An interfaces "exports" are to be imported by a module using the interface and the interfaces "imports" are to be exported by the module using the interface. Since imports are to be provided by the module using the interface the have no module namespace associated with them.  
+
+```
+(interface $a
+  ;; the module using this interface must export "memory"
+  (import "memory" (memory)) 
+  (global $a (export "a_global") (mut i32) (i32.const -2))
+  (func (export "a_func"))
+)
+```
+
+### Results
+watidl can have multiple results and results can be labeled
+```
+(func (export "multi-vale")
+  (result $result1 u64)
+  (result $result2 f32)
+)
+```
+
+### Object Oriented
+An interface is also a first class type
+
+```
+(interface $b)
+
+(interface $a
+  (func (export "factory")
+    (result $the_result $b)
```

Comment:
Could you describe what this means in more detail? Would each instance of an interface have its own copy of the exported global variables declared in an interface?

Reply:
> Would each instance of an interface have its own copy of the exported global variables declared in an interface?

thats a good question, and it show that might be some inconsistency here. i'll think about this more